cardano-crypto-class/src/Cardano/Crypto/PinnedSizedBytes.hs

```diff
@@ -202,28 +211,134 @@ instance KnownNat n => Storable (PinnedSizedBytes n) where
+-- | As 'psbUseAsCPtr', but also gives the function argument the size we are
+-- allowed to use as a 'CSize'.
+--
+-- This is mostly boilerplate removal, as it is quite common for C APIs to take
+-- a combination of a pointer to some data and its length. A possible use case
+-- (and one we run into) is where we know that we can expect a certain data
+-- length (using 'PinnedSizedBytes' as its representation), but the C API allows
+-- any length we like, provided we give the right argument to indicate this.
+-- Therefore, having a helper like this one allows us to avoid having to
+-- manually 'natVal' a 'Proxy', as well as ensuring we don't get mismatches
+-- accidentally.
+--
+-- The same caveats apply to the use of this function as to the use of
+-- 'psbUseAsCPtr'.
+psbUseAsCPtrLen :: 
+  forall (n :: Nat) (r :: Type) . 
+  (KnownNat n) => 
+  PinnedSizedBytes n ->
+  (Ptr Word8 -> CSize -> IO r) -> 
+  IO r
+psbUseAsCPtrLen (PSB ba) f = do
+  let len :: CSize = fromIntegral . natVal $ Proxy @n
+  runAndTouch ba (`f` len)
+
+-- | As 'psbUseAsCPtr', but does not \'forget\' the size.
+--
+-- The same caveats apply to this use of this function as to the use of
+-- 'psbUseAsCPtr'.
+psbUseAsSizedPtr :: 
+  forall (n :: Nat) (r :: Type) . 
+  PinnedSizedBytes n -> 
+  (SizedPtr n -> IO r) -> 
+  IO r
 psbUseAsSizedPtr (PSB ba) k = do
     r <- k (SizedPtr $ castPtr $ byteArrayContents ba)
     r <$ touch ba
 
-psbCreate :: forall n. KnownNat n => (Ptr Word8 -> IO ()) -> IO (PinnedSizedBytes n)
-psbCreate k = do
-    let size :: Int
-        size = fromInteger (natVal (Proxy :: Proxy n))
-    mba <- newPinnedByteArray size
-    k (mutableByteArrayContents mba)
-    arr <- unsafeFreezeByteArray mba
-    return (PSB arr)
+-- | As 'psbCreateResult', but presumes that no useful value is produced: that
+-- is, the function argument is run only for its side effects.
+psbCreate :: 
+  forall (n :: Nat) . 
+  (KnownNat n) => 
+  (Ptr Word8 -> IO ()) -> 
+  IO (PinnedSizedBytes n)
+psbCreate f = fst <$> psbCreateResult f
+
+-- | As 'psbCreateResultLen', but presumes that no useful value is produced:
+-- that is, the function argument is run only for its side effects.
+psbCreateLen :: 
+  forall (n :: Nat) . 
+  (KnownNat n) => 
+  (Ptr Word8 -> CSize -> IO ()) -> 
+  IO (PinnedSizedBytes n)
+psbCreateLen f = fst <$> psbCreateResultLen f
 
-psbCreateSized :: forall n. KnownNat n => (SizedPtr n -> IO ()) -> IO (PinnedSizedBytes n)
+-- | Given an \'initialization action\', which also produces some result, allocate
+-- new pinned memory of the specified size, perform the action, then return the
+-- result together with the initialized pinned memory (as a 'PinnedSizedBytes').
+--
+-- = Note
+--
+-- It is essential that @r@ is not the 'Ptr' given to the function argument:
+-- returning this would be extremely unsafe, as it would break referential
+-- transparency guarantees by allowing us to alias supposedly immutable memory.
```

Comment:
It is also unsafe to read from such a `Ptr` because the `PinnedSizedBytes` could be GCed by the time you peek into it.
So something like
```haskell
psbCreateResult pure >>= \(_, ptr) -> threadDelay 1000000 >> peek ptr
```
could easily segfault or more likely refer to some other totally unrelated memory buffer and thus would pose a potential security vulnerability